File: 70-climbing-stairs/70-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        # O(n) time & O(n) space
        
        num_steps = [0] * (n + 1)  # Number of steps to reach i
        
        num_steps[0] = 1
        num_steps[1] = 1
        
        for i in range(2, n + 1):
            num_steps[i] = num_steps[i - 1] + num_steps[i - 2]
            
        return num_steps[n]

File: 70-climbing-stairs/70-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        return climbStairsTopDownClient(n);
    }
    
    private int climbStairsTopDownClient(int n) {
        int[] countDP = new int[n + 1];
        return climbStairsTopDown(n, countDP);
    }
    
    private int climbStairsTopDown(int n, int[] countDP) {
        // Top down / Memoization
        // O(n) time & O(n) space
        
        // Base case
        if (n < 0) return 0;
        if (n == 0) return 1;
        
        // Already computed
        if (countDP[n] != 0) return countDP[n];
        
        // Recursive case
        int stepX = climbStairsTopDown(n - 1, countDP);  // Go down by 1 step
        int stepY = climbStairsTopDown(n - 2, countDP);  // Go down by 2 steps
        
        countDP[n] = stepX + stepY;
        
        return countDP[n];
    }
}